### PR TITLE
added rwvfs.Union

### PR DESCRIPTION
**Pull request by neelance**

### CODE REVIEW

File: union.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	for _, fs := range fileSystems {
+		ns.Bind("/", fs, "/", vfs.BindAfter)
+	}
+	return &unionFs{
+		NameSpace:   ns,
+		fileSystems: fileSystems,
+	}
+}
+
+type unionFs struct {
```
Comment (sqs) [inline union.go]: for consistency, `unionFS` not `unionFs`

File: vfs_test.go [hunk 2]
```diff
@@ -608,3 +609,34 @@ func TestOS_ReadLink_ErrOutsideRoot(t *testing.T) {
...
+
+	testCreate(t, u, "test", m1)
+	testCreate(t, u, "foo/test", m1)
+	testCreate(t, u, "bar/test", m2)
+}
+
+func testCreate(t *testing.T, u rwvfs.FileSystem, path string, m map[string]string) {
+	w, err := u.Create(path)
+	if err != nil {
+		panic(err)
```
Comment (sqs) [inline vfs_test.go]: Why not `t.Fatal(err)`?

File: union.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	vfs.NameSpace
+	fileSystems []FileSystem
+}
+
+func (fs *unionFs) resolve(path string) FileSystem {
+	parent := pathpkg.Dir(path)
+	for _, fs2 := range fs.fileSystems {
+		if _, err := fs2.Stat(parent); err == nil {
```
Comment (sqs) [inline union.go]: Seems to me that this should check for existence, not just for lack of error. What if the first filesystem returns an error that either (A) indicates that an unrelated error occurred and that the dir may or may not exist, or (B) indicates that the file exists but some other error occurred?

`man fstat` shows a bunch of possible errors:

```
ERRORS
       EACCES Search permission is denied for one of the directories in the path prefix of path.   (See
              also path_resolution(7).)

       EBADF  fd is bad.

       EFAULT Bad address.

       ELOOP  Too many symbolic links encountered while traversing the path.

       ENAMETOOLONG
              path is too long.

       ENOENT A component of path does not exist, or path is an empty string.

       ENOMEM Out of memory (i.e., kernel memory).

       ENOTDIR
              A component of the path prefix of path is not a directory.

       EOVERFLOW
              path  or fd refers to a file whose size, inode number, or number of blocks cannot be rep‐
              resented in, respectively, the types off_t, ino_t, or blkcnt_t.   This  error  can  occur
              when,  for  example,  an  application  compiled on a 32-bit platform without -D_FILE_OFF‐
              SET_BITS=64 calls stat() on a file whose size exceeds (1<<31)-1 bytes.
```

Not all of these indicate that the file doesn't exist. Obviously those don't map 1-to-1 to errors that `Stat` would return, but they're a good starting point.

I am not sure, but I think this might be better behavior:

```go
func (fs *unionFS) resolve(path string) (FileSystem, error) {
  // ...
  if _, err := fs2.Stat(parent); err == nil {
    return fs2, nil
  } else if !os.IsNotExist(err) {
    return nil, err
  }
// ...
}
```

This seems pedantic, but this is such a low-level library that it's worth it IMO.

File: union.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+			return fs2
+		}
+	}
+	return fs.fileSystems[0]
+}
+
+// Create creates the named file, truncating it if it already exists.
+func (fs *unionFs) Create(path string) (io.WriteCloser, error) {
+	return fs.resolve(path).Create(path)
```
Comment (sqs) [inline union.go]: A lock should be held during both the calls to `fs.resolve(path)` and `(FileSystem).{Create,Mkdir,Remove}`. That's because the latter operation affects the result of the former, so you have [TOCTTOU](http://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) race conditions here.

This might really hurt performance, unfortunately. I think that's highly likely. In that case, I think you can just:

* Try to think if there are any really bad bugs (panics, etc.) that could occur.
* If not, then just add a comment to the Union func and/or unionFS type explaining the race condition and why we're punting on solving it.
* If so, then we probably do need to fix it.

File: union.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	"io"
+	pathpkg "path"
+
+	"golang.org/x/tools/godoc/vfs"
+)
+
+// Union returns a new FileSystem which is the union of the provided file systems.
+func Union(fileSystems ...FileSystem) FileSystem {
+	if len(fileSystems) == 0 {
```
Comment (sqs) [inline union.go]: This is just personal taste, but I find this clearer:

```go
switch len(fileSystems) {
  case 0:
    return Map(nil)
  case 1:
    return fileSystems[0]
}
```

File: union.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		return Map(nil)
```
Comment (sqs) [inline union.go]: I think that returning `Map(nil)` is not the ideal behavior here for a few reasons:

* I believe it'll cause a panic when you try to perform an FS operation, because `Map` doesn't make an empty map if its argument is nil. That is solvable, but...
* At a conceptual level, what is the union of 0 filesystems? I would argue it is a meaningless concept and should be `nil`. Here you *create* a new real filesystem and return it.

So I think it would be better to mention in the documentation that if no filesystems are passed to `Union`, it returns `nil` (or panics)--and implement it that way, too.

File: union.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	return &unionFs{
+		NameSpace:   ns,
+		fileSystems: fileSystems,
+	}
+}
+
+type unionFs struct {
+	vfs.NameSpace
+	fileSystems []FileSystem
+}
```
Comment (sqs) [inline union.go]: Add a `(unionFS).String() string` method to unionFS, or else it'll fall back to `vfs.Namespace`'s String method (which will be confusing).

File: vfs_test.go (hunk 2 above)
Comment (neelance) [inline vfs_test.go]: Missed that, thx. Will fix.

File: union.go (hunk 3 above)
Comment (neelance) [inline union.go]: What do you think of the following?
```go
func (fs *unionFs) resolve(path string) FileSystem {
	parent := pathpkg.Dir(path)
	for _, fs2 := range fs.fileSystems {
		if _, err := fs2.Stat(parent); err != nil && os.IsNotExist(err) {
			continue;
		}
		return fs2
	}
	return fs.fileSystems[0]
}
```
The reason for not returning an error here is that I want to return an error for `path` itself, not for its parent directory.

File: union.go (hunk 4 above)
Comment (neelance) [inline union.go]: I saw that race condition, but you can't fix it with a mutex here, since the underlying file systems may also be changed from somewhere else, even other processes in case of OS file systems. Synchronization (if possible) needs to be done at a higher level, e.g. the consumer of this API. Maybe the documentation should state that explicitly?

File: union.go (hunk 5 above)
Comment (neelance) [inline union.go]: For me a `switch` means "code paths of equal importance", often seen with enum constants. Here we have two special cases that are not related to each other. You may even remove the check for `1` and it would still work fine. That's why I intentionally prefer two `if` statements.

File: union.go (hunk 6 above)
Comment (neelance) [inline union.go]: Is it really meaningless? I'd say the union of zero filesystems is an empty filesystem, just like the union of zero sets is the empty set (and a filesystem is a set of files). Of course write access should fail. Maybe something with `rwvfs.ReadOnly`?

File: union.go (hunk 7 above)
Comment (neelance) [inline union.go]: Good point, will fix.

File: union.go (hunk 1 above)
Comment (neelance) [inline union.go]: :+1:

**Comment by neelance** [comment]
@sqs Do you have an idea why Travis fails? And should it be upgraded to Go 1.4?

**Comment by shurcooL** [comment]
It's building on `go version go1.1 linux/amd64` and tip. That is way too outdated. I think you should just replace it with Go 1.4 (and if you want, tip, but make it an allowed failure).

However, it's failing because of a build error:

```
# github.com/sourcegraph/rwvfs_test
./vfs_test.go:84: undefined: rwvfs.Union
./vfs_test.go:616: undefined: rwvfs.Union
FAIL	github.com/sourcegraph/rwvfs [build failed]
```

File: union.go (hunk 5 above)
Comment (sqs) [inline union.go]: 👍🏾

File: union.go (hunk 4 above)
Comment (sqs) [inline union.go]: 👍🏿

File: union.go (hunk 6 above)
Comment (sqs) [inline union.go]: 👍

**Comment by shurcooL** [comment]
Okay, it's failing because the import part is `sourcegraph.com/sourcegraph/rwvfs` (in `vfs_test.go`), but the Travis script doesn't take that into account (it's cloned into GOPATH as `github.com/sourcegraph/rwvfs`).

You can copy the Travis script from https://github.com/sourcegraph/go-diff/blob/master/.travis.yml, that will fix it.

```

File: vfs_test.go [hunk 8]
```diff
@@ -608,3 +609,34 @@ func TestOS_ReadLink_ErrOutsideRoot(t *testing.T) {
...
+	}
+	w.Write([]byte("test"))
+	w.Close()
+
+	if got, want := m[path], "test"; got != want {
+		t.Errorf(`Create: got %d, want %d`, got, want)
```
Comment (shurcooL) [inline vfs_test.go]: An incorrect verb `%d` is being used here for type `string`. Either `%s` or the more general `%v` should be used instead.

File: union.go (hunk 1 above)
Comment (neelance) [inline union.go]: Fixed by b6abb77.

File: vfs_test.go (hunk 2 above)
Comment (neelance) [inline vfs_test.go]: Fixed by a01ef80.

File: union.go (hunk 3 above)
Comment (neelance) [inline union.go]: Pushed my proposed solution with a42deb4.

File: union.go (hunk 4 above)
Comment (neelance) [inline union.go]: Added documentation with dfd6c63.

File: union.go (hunk 6 above)
Comment (neelance) [inline union.go]: Fixed implementation and added test with 51e2380.

File: union.go (hunk 7 above)
Comment (neelance) [inline union.go]: Fixed by 51e2380.

**Comment by neelance** [comment]
I've addressed all comments. I will squash all `Union` commits into one before merging it into master. LGTM?

**Comment by sqs** [comment]
I won't have time to check it this week - @shurcool, can you take a look? I am sure it is good, though. :)

**Comment by shurcooL** [comment]
Looking through it now.

File: union.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+	pathpkg "path"
+
+	"golang.org/x/tools/godoc/vfs"
+)
+
+// Union returns a new FileSystem which is the union of the provided file systems.
```
Comment (shurcooL) [inline union.go]: How precisely do you want it to be a "union"? Is it okay if some features are not supported?

For example, is it a union of distinct filesystems, or is overlap between filesystems allowed? If allowed, does it gracefully merge contents of folders with same path?

I ask because I know it currently does not handle every single possible scenario, but perhaps that's okay. It would be nice to see the limitations documented though, as that would clear up this question. On the other hand, if having correct union of overlapping filesystems is the goal, that should be fixed (I can provide a failing test case).

Comment (shurcooL) [inline union.go]: Also, is it okay/desired that there is special handling of .go files inherited from `"golang.org/x/tools/godoc/vfs.NameSpace`, since this is a wrapper around that?

See https://godoc.org/golang.org/x/tools/godoc/vfs#NameSpace.ReadDir and http://gotools.org/golang.org/x/tools/godoc/vfs#namespace-go-L318-L330.

**Comment by shurcooL** [comment]
Everything else looks okay to me, but without knowing the exact goal/scope of `Union`, I'm not sure how thoroughly I should look through its behavior.

**Comment by shurcooL** [comment]
From context, I am guessing that the behavior is as you expect it. If that's the case, I think it'll be good to add some docs on the scope and limitations of this `Union` function. Otherwise, by describing it as a very general "filesystem union" without stating limitations, we're setting ourselves up for multiple open issues/bug reports, and I'm not sure if our goal is to resolve those (if it is, then it's good now).

By setting the scope smaller (via documentation), that problem goes away.

**Comment by shurcooL** [comment]
That addresses my points. LGTM.

**Comment by shurcooL** [comment]
Closed via 486352b6c67dee55205b70de6cecf7e9ef1e696d.
